crates/oxc_linter/src/rules/no_constant_condition.rs

```diff
@@ -0,0 +1,412 @@
+use oxc_ast::{ast::Expression, AstKind, Span};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+
+use crate::{ast_util::IsConstant, context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(no-constant-condition): Unexpected constant condition")]
+#[diagnostic(severity(warning), help("Constant expression as a test condition is not allowed"))]
+struct NoConstantConditionDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoConstantCondition {
+    _check_loops: bool,
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallow constant expressions in conditions
+    ///
+    /// ### Why is this bad?
+    ///
+    /// A constant expression (for example, a literal) as a test condition might be a typo or development trigger for a specific behavior.
+    ///
+    /// ### Example
+    ///
+    /// ```javascript
+    /// if (false) {
+    ///    doSomethingUnfinished();
+    /// }
+    /// ```
+    NoConstantCondition,
+    nursery
+);
+
+impl Rule for NoConstantCondition {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let obj = value.get(0);
+
+        Self {
+            _check_loops: obj
+                .and_then(|v| v.get("checkLoops"))
+                .and_then(serde_json::Value::as_bool)
+                .unwrap_or_default(),
+        }
+    }
+
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        match node.get().kind() {
+            AstKind::IfStatement(if_stmt) => {
+                if if_stmt.test.is_constant(true, ctx) {
+                    diagnose_constant_expr(&if_stmt.test, ctx);
+                }
+            }
+            AstKind::ConditionalExpression(cond_expr) => {
+                if cond_expr.test.is_constant(true, ctx) {
+                    diagnose_constant_expr(&cond_expr.test, ctx);
+                }
+            }
+            _ => {}
+        }
+    }
+}
+
+fn diagnose_constant_expr(expr: &Expression, ctx: &LintContext) {
+    let diagnose = |span: Span| {
+        ctx.diagnostic(NoConstantConditionDiagnostic(span));
```

Comment:
you can get the span of the expr by `expr.span()`, just import the `GetSpan` trait from oxc_ast